docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+
+    box Management Cluster
+        participant apiserver as kube-api server
+        participant capi as CP/MD controller
+        participant mach as Machine Controller
+        participant hook as External updater
+    end
+
+
+    Operator->>+apiserver: make changes to KCP <br>(ExternalUpdate strategy)
+    apiserver->>+capi: Notify changes
+    apiserver->>-Operator: OK
+    loop For all External Updaters
+        capi->>+hook: Can update?
+        hook->>capi: Supported changes
+    end
+    capi->>capi: Define Update Plan
+    loop For all machines
+        capi->>apiserver: Mark Machine with Update Plan
+        apiserver->>mach: Notify changes
+        loop For all updaters in plan
+            mach->>hook: Run updater
+        end
+        mach->>apiserver: Mark Machine as updated
+    end
+```
+
+When configured, external updates will, roughly, follow these steps:
+1. CP/MD Controller: detect an update is required.
+2. CP/MD Controller: query defined update extensions, and based on the subset of changes each of them supports, defines the final update plan.
+3. CP/MD Controller: mark machines for update.
+4. Machine Controller: invoke all the updaters included in the plan, sequentially, one by one.
+5. Machine Controller: make machine is updated.
+
+The following sections dive deep into these steps, zooming in into the different component interactions and defining how the main error cases are handled.
+
+### Defining the Update Plan
+
+```mermaid
+sequenceDiagram
+    participant Operator
+
+    box Management Cluster
+        participant apiserver as kube-api server
+        participant capi as CP/MD Controller
+        participant hook as External updater 1
+        participant hook2 as External updater 2
+    end
+
+    Operator->>+apiserver: make changes to CP/MD <br>(ExternalUpdate strategy)
+    apiserver->>+capi: Notify changes
+    apiserver->>-Operator: OK
+    capi->>+hook: Can update?
+    hook->>capi: Set of changes
+    capi->>+hook2: Can update?
+    hook2->>capi: Set of changes
+    alt all changes covered?
+        capi->>apiserver: Define Update Plan
+    else
+        alt fallback strategy?
+            capi->>apiserver: Re-create machines
+        else
+            capi->>apiserver: Marked update as failed
+        end
+    end
+```
+
+Both `KCP` and `MachineDeployment` controllers follow a similar pattern around updates, they first detect if an update is required and then based on the configured strategy follow the appropiate update logic (note that today there is only one valid strategy, `RollingUpdate`).
+
+With `ExternalUpdate` strategy, CAPI controllers will compute the set of desired changes and iterate over the registered external updaters, requesting through the Runtime Hook the set of changes each updater can handle. The changes supported by an updater can be the complete set of desired changes, a subset of them or an empty set, signaling it cannot handle any of the desired changes.
```

Comment:
That's an interesting point. I'm not against representing external updates as a subtype of rolling update strategy. You are right that we what we are proposing here, CAPI is following a rolling update process except it delegates the machine update instead of replacing the machine by itself. But capi orchestrates the rolling process.

As long as we can represent the fallback as optional, I'm ok with this if folks think it makes more sense.